constraints/src/main/java/gov/nasa/jpl/aerie/constraints/time/Spans.java

```diff
@@ -13,76 +13,76 @@
 /**
  * A collection of intervals that can overlap.
  */
-public class Spans implements Iterable<Window> {
-  private final List<Window> windows;
+public class Spans implements Iterable<Interval> {
+  private final List<Interval> intervals;
 
   public Spans() {
-    this.windows = new ArrayList<>();
+    this.intervals = new ArrayList<>();
   }
 
-  public Spans(final ArrayList<Window> windows) {
-    windows.removeIf(Window::isEmpty);
-    this.windows = windows;
+  public Spans(final ArrayList<Interval> intervals) {
+    intervals.removeIf(Interval::isEmpty);
+    this.intervals = intervals;
   }
 
-  public Spans(final Iterable<Window> iter) {
-    this.windows = StreamSupport.stream(iter.spliterator(), false).filter($ -> !$.isEmpty()).toList();
+  public Spans(final Iterable<Interval> iter) {
+    this.intervals = StreamSupport.stream(iter.spliterator(), false).filter($ -> !$.isEmpty()).toList();
   }
 
-  public Spans(final Window... windows) {
-    this.windows = Arrays.stream(windows).filter($ -> !$.isEmpty()).toList();
+  public Spans(final Interval... intervals) {
+    this.intervals = Arrays.stream(intervals).filter($ -> !$.isEmpty()).toList();
   }
 
   public Windows intoWindows() {
-    return new Windows(this.windows);
+    return Windows.definedEverywhere(this.intervals, true);
```

Comment:
Do you know if anything depends on the produced windows being `false` and not merely `null` outside the spans? I can see the reasoning for it, but it still makes me uncomfortable.

In particular, I see that it's only used by a constraint language operator, `WindowsFromSpans`, so I imagine this is relevant to concrete user stories.

On the other hand, `WindowsFromSpans` already has access to the bounds of the simulation results, so it doesn't seem unreasonable to use those bounds rather than `Interval.FOREVER`.

Reply:
Converting Spans to Windows should always produce `false` outside of the spans, not `null`. Otherwise it would litter constraints evaluation with large gaps.

> Do you know if anything depends on the produced windows being false and not merely null outside the spans?

everything in scheduler

Reply:
Would it be reasonable to use the bounds from `SimulationResults` instead of `definedEverywhere` and its internal use of `Interval.FOREVER` here, since they're already immediately at hand?

Reply:
I don't think using `Interval.FOREVER` is a problem at all, and I don't think this conversion should be limited by the simulation bounds. It wasn't limited before. If you shift an interval outside of the plan, it should not be truncated until the consumer of the Spans or Windows object decides to do so.

If you don't like `Interval.FOREVER`, fix the thing you actually have an issue with. Don't limit everything else just because we have a flawed representation of forever.

Reply:
> I don't think using `Interval.FOREVER` is a problem at all, and I don't think this conversion should be limited by the simulation bounds. It wasn't limited before. If you shift an interval outside of the plan, it should not be truncated until the consumer of the Spans or Windows object decides to do so.

Hm. Fair point. 🤔

I'm starting to think that `IntervalMap` having bounded support is not essential. Actually, I think what depended on bounded support is just the internal representation of `IntervalSet` (and now `IntervalMap`) as a list of intervals. I didn't want to add values for positive and negative infinity to `Duration`, and an `Interval` is *morally* an ordered pair of `Duration`s, so bounded support fell out of that design constraint.

In [my last prototype of `IntervalMap`](https://gist.github.com/Twisol/88a8b2bf9e71681e74c1b4160ee9a03e), I used a representation as a list of *transitions*, which is also exactly the representation used for profiles in the database. Some of the invariants of that implementation only exist to retain bounded support; if we dropped that requirement, we need only add a field `V initialValue` to represent the value taken by the timeline infinitely long before the first transition. The last transition (if any) already defines the value carried infinitely long into the future. This doesn't even require adding infinite points to `Duration` or `Interval`.

The problem arises when iterating over the `IntervalMap`, because we still have no way to represent the period of time from negative infinity up to the first transition (or from the last transition to positive infinity). We could change the iteration behavior from intervals to transitions, but it's not super clear to me how that would affect client code. We could also just add points at infinity to `Interval`; this would let us keep the current representation of `IntervalMap`, but we still have to accommodate the new semantics in the DB and UI representations. (The current representation of `IntervalMap` is [wasteful](https://github.com/NASA-AMMOS/aerie/discussions/248#discussioncomment-3184164) for gapless profiles, since every pair of adjacent intervals repeats their shared endpoint, so I'd still like to move away from it eventually.)

I still don't want to change `Duration`, because the primary owner of `Duration`, the simulation system, has no need for infinities. Plus, no transition ever occurs at infinity; if we continued with the `Interval`-based internal representation, if we have N intervals and hence 2N endpoints, at most *two* of them will be infinite. These facts suggest that infinite intervals are just a totally different concern from `Duration` altogether.

> If you don't like `Interval.FOREVER`, fix the thing you actually have an issue with. Don't limit everything else just because we have a flawed representation of forever.

This comes off as a lot more aggressive than I think you meant it to be. 😅

Reply:
All this to say, I'm still not a fan of `Interval.FOREVER` and `definedEverywhere`, because they're fundamentally not on solid semantic footing. But now, instead of trying to excise them from a system too impoverished to support them, I agree that they're sensible concepts that we should support directly.

Reply:
I think I've convinced myself that adding infinities to `Interval` is a well-founded way to go, and should work regardless of the implementation we use. I think it works out most simply with the transition-based representation; I think we can get away with adding just one `INFINITY` value whose sign is implied by whether it appears as the start point or the end point. The transition-based representation doesn't store infinities in the first place, so there's no confusion about comparing a duration to `INFINITY` and not knowing whether it's positive or negative. It's not obvious to me whether that's true for the interval-based representation. The easy fix is to add two separate signed infinities, but then you have to exclude nonsense intervals like `+INF, -INF`, and *these things aren't real anyway*.

```java
public final class Interval implements Comparable<Interval> {
  public sealed interface Bound {
    record Infinity() implements Bound {}
    record Point(Duration time, Inclusivity inclusivity) implements Bound {}
  }

  public final Bound start;
  public final Bound end;
}
```

Reply:
I think I have resolved a few interconnected comment threads now, so I'm just commenting here.

I've re-implemented `Windows` as delegating to IntervalMap instead of inheriting, and non-mutating methods. We are both right with regard to usability and maintenance. The interface is smoother to use and more consistent in its behavior, and also the giant block of delegating methods is already a pain to keep up to date.

I've removed `IntervalAlgebra.horizon`, and converted `IntervalAlgebra` to be fully static. Bounding the resulting `Windows` is now the responsibility of the Windows consumer, explicitly through the `select` method.

I agree that the transitions would be cleaner with respect to infinities, but that refactor would be gargantuan (I'd end up re-doing the entire PR at minimum). I also agree that allowing Interval to be unbounded is also good solution, but due to the database issue I don't think I should try to do that in this PR.

Reply:
> I agree that the transitions would be cleaner with respect to infinities, but that refactor would be gargantuan (I'd end up re-doing the entire PR at minimum). I also agree that allowing Interval to be unbounded is also good solution, but due to the database issue I don't think I should try to do that in this PR.

Yes, definitely agree. I think it's fine as long as we know where we can go with this in the future.

Reply:
> I've re-implemented `Windows` as delegating to IntervalMap instead of inheriting, and non-mutating methods. We are both right with regard to usability and maintenance. The interface is smoother to use and more consistent in its behavior, and also the giant block of delegating methods is already a pain to keep up to date.

I think we could go all-in on making `Windows` a fully static class, and have its static methods accept and return `IntervalMap<Boolean>`. That would remove a lot of the delegation issue, but would be a little less idiomatic re: fluent interfaces. But I don't think we need to tackle that in this PR either.

Reply:
I really want to just

```rust
struct IntervalMap<V>(Vec<Segment<V>>);

impl<V> IntervalMap<V> {
    fn map(...) -> ...
}

impl IntervalMap<bool> {
    fn and(...) -> ...
    fn or(...) -> ...
}
```

but alas

Reply:
If you can get buy-in on [Manifold](http://manifold.systems/), we can get extension method syntax sugar. 👀

Reply:
(though, honestly, Manifold moves just slowly enough relative to Java releases that I wouldn't be comfortable pinning to Manifold's supported versions. Dangit.)